### PR TITLE
Pre-decompressed DNS record RData that may contain compression pointers

### LINKED ISSUE #8962: netty-resolver-dns "Got bad packet: bad label type"
I have built a simple DNS proxy server using the Netty DNS resolver library. [See the code here.](https://github.com/nzhenry/dns-proxy)

It works for most queries but not all. For example, when I try to query the A records for "www.apple.com" using nslookup or dig, it fails intermittently with this error:
```;; Got bad packet: bad label type
305 bytes
ca 20 80 00 00 01 00 04 00 00 00 01 03 77 77 77          .............www
05 61 70 70 6c 65 03 63 6f 6d 00 00 01 00 01 03          .apple.com......
77 77 77 05 61 70 70 6c 65 03 63 6f 6d 00 00 05          www.apple.com...
00 01 00 00 00 37 00 1b 03 77 77 77 05 61 70 70          .....7...www.app
6c 65 03 63 6f 6d 07 65 64 67 65 6b 65 79 03 6e          le.com.edgekey.n
65 74 00 03 77 77 77 05 61 70 70 6c 65 03 63 6f          et..www.apple.co
6d 07 65 64 67 65 6b 65 79 03 6e 65 74 00 00 05          m.edgekey.net...
00 01 00 00 30 2b 00 2f 03 77 77 77 05 61 70 70          ....0+./.www.app
6c 65 03 63 6f 6d 07 65 64 67 65 6b 65 79 03 6e          le.com.edgekey.n
65 74 0b 67 6c 6f 62 61 6c 72 65 64 69 72 06 61          et.globalredir.a
6b 61 64 6e 73 c0 41 03 77 77 77 05 61 70 70 6c          kadns.A.www.appl
65 03 63 6f 6d 07 65 64 67 65 6b 65 79 03 6e 65          e.com.edgekey.ne
74 0b 67 6c 6f 62 61 6c 72 65 64 69 72 06 61 6b          t.globalredir.ak
61 64 6e 73 03 6e 65 74 00 00 05 00 01 00 00 0c          adns.net........
b7 00 19 05 65 36 38 35 38 05 64 73 63 65 39 0a          ....e6858.dsce9.
61 6b 61 6d 61 69 65 64 67 65 c0 41 05 65 36 38          akamaiedge.A.e68
35 38 05 64 73 63 65 39 0a 61 6b 61 6d 61 69 65          58.dsce9.akamaie
64 67 65 03 6e 65 74 00 00 01 00 01 00 00 00 11          dge.net.........
00 04 68 5f a0 7e 00 00 29 10 00 00 00 00 00 00          ..h_....).......
00                                                       .
```
When it fails the debug log looks like this:
```DefaultDnsQuestion(www.apple.com. IN A)
DefaultDnsRawRecord(www.apple.com. 885 IN CNAME 27B)
DefaultDnsRawRecord(www.apple.com.edgekey.net. 10239 IN CNAME 47B)
DefaultDnsRawRecord(www.apple.com.edgekey.net.globalredir.akadns.net. 441 IN CNAME 25B)
DefaultDnsRawRecord(e6858.dsce9.akamaiedge.net. 9 IN A 4B)] RECEIVED: [[[id: 0xc9dafb10], 36804, /192.168.1.1:53, DatagramDnsResponse(from: /192.168.1.1:53, to: /0:0:0:0:0:0:0:0:51678, 36804, QUERY(0), NoError(0), RD RA)
DefaultDnsQuestion(www.apple.com. IN A)
DefaultDnsRawRecord(www.apple.com. 885 IN CNAME 27B)
DefaultDnsRawRecord(www.apple.com.edgekey.net. 10239 IN CNAME 47B)
DefaultDnsRawRecord(www.apple.com.edgekey.net.globalredir.akadns.net. 441 IN CNAME 25B)
DefaultDnsRawRecord(e6858.dsce9.akamaiedge.net. 9 IN A 4B)]: {}], {}
```
When it succeeds the log looks like this:
```DefaultDnsQuestion(www.apple.com. IN A)
DefaultDnsRawRecord(www.apple.com. 881 IN CNAME 27B)
DefaultDnsRawRecord(www.apple.com.edgekey.net. 10235 IN CNAME 50B)
DefaultDnsRawRecord(www.apple.com.edgekey.net.globalredir.akadns.net. 437 IN CNAME 28B)
DefaultDnsRawRecord(e6858.dsce9.akamaiedge.net. 5 IN A 4B)
DefaultDnsRawRecord(OPT flags:0 udp:4096 0B)] RECEIVED: [[[id: 0xc9dafb10], 37086, /192.168.1.1:53, DatagramDnsResponse(from: /192.168.1.1:53, to: /0:0:0:0:0:0:0:0:51678, 37086, QUERY(0), NoError(0), RD RA)
DefaultDnsQuestion(www.apple.com. IN A)
DefaultDnsRawRecord(www.apple.com. 881 IN CNAME 27B)
DefaultDnsRawRecord(www.apple.com.edgekey.net. 10235 IN CNAME 50B)
DefaultDnsRawRecord(www.apple.com.edgekey.net.globalredir.akadns.net. 437 IN CNAME 28B)
DefaultDnsRawRecord(e6858.dsce9.akamaiedge.net. 5 IN A 4B)
DefaultDnsRawRecord(OPT flags:0 udp:4096 0B)]: {}], {}
```
Notice the additional OPT record when it succeeds. This appears to be the key. I just don't know how I can fix the problem when it fails.

I've discovered that if I remove the CNAME records from the response, then the query resolves with no errors. I would like to return a proper copy of the server response including the CNAME records. Any help would be greatly appreciated!

### PR DESCRIPTION
**Pull request by qeesung**
Motivation:

When decoding `DnsRecord`, if the record contains compression pointers, and not all compression pointers are decompressed, but part of the pointers are decompressed. Then when encoding the record, the compressed pointer will point to the wrong location, resulting in **bad label** problem.

Modification:

Pre-decompressed record RData that may contain compression pointers.

Result:

Fixes #8962

### CODE REVIEW

File: codec-dns/src/main/java/io/netty/handler/codec/dns/DnsCodecUtil.java [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufUtil;
+import io.netty.buffer.Unpooled;
+import io.netty.handler.codec.CorruptedFrameException;
+import io.netty.util.CharsetUtil;
+
+import static io.netty.handler.codec.dns.DefaultDnsRecordDecoder.*;
+
+public final class DnsCodecUtil {
```
Comment (normanmaurer) [inline codec-dns/src/main/java/io/netty/handler/codec/dns/DnsCodecUtil.java]: Package private ?

File: codec-dns/src/main/java/io/netty/handler/codec/dns/DnsCodecUtil.java [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    private DnsCodecUtil() {
+        // Util class
+    }
+
+    public static void encodeDomainName(String name, ByteBuf buf) {
```
Comment (normanmaurer) [inline codec-dns/src/main/java/io/netty/handler/codec/dns/DnsCodecUtil.java]: Package private

File: codec-dns/src/main/java/io/netty/handler/codec/dns/DnsCodecUtil.java [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+            }
+
+            buf.writeByte(labelLen);
+            ByteBufUtil.writeAscii(buf, label);
+        }
+
+        buf.writeByte(0); // marks end of name field
+    }
+
+    public static String decodeDomainName(ByteBuf in) {
```
Comment (normanmaurer) [inline codec-dns/src/main/java/io/netty/handler/codec/dns/DnsCodecUtil.java]: Package private

File: codec-dns/src/main/java/io/netty/handler/codec/dns/DnsCodecUtil.java [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+
+        return name.toString();
+    }
+
+    /**
+     * Decompress pointer data.
+     * @param compression comporession data
+     * @return decompressed data
+     */
+    public static ByteBuf decompressDomainName(ByteBuf compression) {
```
Comment (normanmaurer) [inline codec-dns/src/main/java/io/netty/handler/codec/dns/DnsCodecUtil.java]: Package private

File: codec-dns/src/main/java/io/netty/handler/codec/dns/DnsCodecUtil.java (hunk 1 above)
Comment (qeesung) [reply codec-dns/src/main/java/io/netty/handler/codec/dns/DnsCodecUtil.java]: Done!

File: codec-dns/src/main/java/io/netty/handler/codec/dns/DnsCodecUtil.java (hunk 2 above)
Comment (qeesung) [reply codec-dns/src/main/java/io/netty/handler/codec/dns/DnsCodecUtil.java]: Done!

File: codec-dns/src/main/java/io/netty/handler/codec/dns/DnsCodecUtil.java (hunk 3 above)
Comment (qeesung) [reply codec-dns/src/main/java/io/netty/handler/codec/dns/DnsCodecUtil.java]: Done!

File: codec-dns/src/main/java/io/netty/handler/codec/dns/DnsCodecUtil.java (hunk 4 above)
Comment (qeesung) [reply codec-dns/src/main/java/io/netty/handler/codec/dns/DnsCodecUtil.java]: Done

**Comment by normanmaurer** [comment]
@netty-bot test this please

File: codec-dns/src/main/java/io/netty/handler/codec/dns/DnsCodecUtil.java [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+    }
+
+    /**
+     * Decompress pointer data.
+     * @param compression comporession data
+     * @return decompressed data
+     */
+    static ByteBuf decompressDomainName(ByteBuf compression) {
+        String domainName = decodeDomainName(compression);
+        ByteBuf result = Unpooled.buffer(domainName.length() << 1);
```
Comment (normanmaurer) [inline codec-dns/src/main/java/io/netty/handler/codec/dns/DnsCodecUtil.java]: use `compression.alloc().buffer(...)` to ensure the same allocator is used (which may be a pooled allocated)

Comment (qeesung) [reply codec-dns/src/main/java/io/netty/handler/codec/dns/DnsCodecUtil.java]: Done

File: codec-dns/src/test/java/io/netty/handler/codec/dns/DefaultDnsRecordDecoderTest.java [hunk 6]
```diff
@@ -89,6 +90,81 @@ public void testDecodePtrRecord() throws Exception {
...
+    @Test
+    public void testdecompressCompressPointer() {
+        byte[] compressionPointer = {
+                5, 'n', 'e', 't', 't', 'y', 2, 'i', 'o', 0,
+                (byte) 0xC0, 0
+        };
+        ByteBuf buffer = Unpooled.wrappedBuffer(compressionPointer);
+        ByteBuf uncomporessed = null;
+        try {
+            uncomporessed = DnsCodecUtil.decompressDomainName(buffer.duplicate().setIndex(10, 12));
```
Comment (normanmaurer) [inline codec-dns/src/test/java/io/netty/handler/codec/dns/DefaultDnsRecordDecoderTest.java]: nit: typo... should be "uncompressed"

File: codec-dns/src/test/java/io/netty/handler/codec/dns/DefaultDnsRecordDecoderTest.java [hunk 7]
```diff
@@ -89,6 +90,81 @@ public void testDecodePtrRecord() throws Exception {
...
+    public void testdecompressNestedCompressionPointer() {
+        byte[] nestedCompressionPointer = {
+                6, 'g', 'i', 't', 'h', 'u', 'b', 2, 'i', 'o', 0, // github.io
+                5, 'n', 'e', 't', 't', 'y', (byte) 0xC0, 0, // netty.github.io
+                (byte) 0xC0, 11, // netty.github.io
+        };
+        ByteBuf buffer = Unpooled.wrappedBuffer(nestedCompressionPointer);
+        ByteBuf uncomporessed = null;
+        try {
+            uncomporessed = DnsCodecUtil.decompressDomainName(buffer.duplicate().setIndex(19, 21));
```
Comment (normanmaurer) [inline codec-dns/src/test/java/io/netty/handler/codec/dns/DefaultDnsRecordDecoderTest.java]: nit: should be "uncompressed"

File: codec-dns/src/test/java/io/netty/handler/codec/dns/DefaultDnsRecordDecoderTest.java (hunk 6 above)
Comment (qeesung) [reply codec-dns/src/test/java/io/netty/handler/codec/dns/DefaultDnsRecordDecoderTest.java]: Done

File: codec-dns/src/test/java/io/netty/handler/codec/dns/DefaultDnsRecordDecoderTest.java (hunk 7 above)
Comment (qeesung) [reply codec-dns/src/test/java/io/netty/handler/codec/dns/DefaultDnsRecordDecoderTest.java]: Done

**Comment by normanmaurer** [comment]
@netty-bot test this please

**Comment by normanmaurer** [comment]
@qeesung thanks a lot!

**Pull request merged**